crates/ibc/src/core/ics02_client/events.rs

```diff
@@ -390,3 +390,101 @@ impl From<UpgradeClient> for abci::Event {
         }
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::mock::client_state::client_type as mock_client_type;
+    use crate::mock::header::MockHeader;
+    use ibc_proto::google::protobuf::Any;
+    use tendermint::abci::Event as AbciEvent;
+
+    #[test]
+    fn ibc_to_abci_client_events() {
+        struct Test {
+            kind: IbcEventType,
+            event: AbciEvent,
+            expected_keys: Vec<&'static str>,
+            expected_values: Vec<String>,
+        }
+
+        let client_type = mock_client_type();
+        let client_id = ClientId::new(client_type.clone(), 0).unwrap();
+        let consensus_height = Height::new(0, 10).unwrap();
+        let consensus_heights = vec![Height::new(0, 5).unwrap(), Height::new(0, 7).unwrap()];
+        let header: Any = MockHeader::new(consensus_height).into();
+        let expected_keys = vec![
+            "client_id",
+            "client_type",
+            "consensus_height",
+            "consensus_heights",
+            "header",
+        ];
+        let expected_values = vec![
+            client_id.to_string(),
+            "9999-mock".to_string(),
+            consensus_height.to_string(),
+            consensus_heights
+                .iter()
+                .map(|h| h.to_string())
+                .collect::<Vec<_>>()
+                .join(","),
+            String::from_utf8(hex::encode(header.clone().value)).unwrap(),
+        ];
```

Comment:
This test doesn't give me the confidence that we're constructing the events in the exact same way as ibc-go. We're still using our own objects for the `expected_values`; we want hardcoded strings here too.

The best way to write this test is to 

1. run [ibc-go's `EmitCreateClientEvent()`](https://github.com/cosmos/ibc-go/blob/887cba1658bbb2db46b96d1ae492cbe8095b1871/modules/core/02-client/keeper/events.go#L18) with a given set of arguments
2. print the output events to the console
3. copy/paste the strings to construct `expected_values`
4. Recreate the same arguments in (1) for `client_type`, `client_id`, etc

Repeat for the connection/channel events tests.

Make sure to use the `main` branch for ibc-go; the `"consensus_heights"` attribute doesn't exist yet in 5.0.1, and I made the mistake of adding it to ibc-rs.

Reply:
Should be fine now https://github.com/cosmos/ibc-rs/pull/363/commits/dc20ce72c523360cd693a2af000146f728cbbc5c
Just an issue I came across... Is it ok to have a default implementation for the `ClientId` ([here](https://github.com/cosmos/ibc-rs/blob/ef52941e29aa2759e2861cdae8a7398b114cc5f2/crates/ibc/src/core/ics24_host/identifier.rs#L241))? 
Therefore, the `core` module becomes chain-specific depending on an `application`. I think we should discard it.

Reply:
of course